core/guard/guard.go

```diff
@@ -0,0 +1,117 @@
+	shard := session.ShardInfo[chunkHash]
+	guardPid, escrowPid, err := getGuardAndEscrowPid(configuration)
+	if err != nil {
+		return nil, err
+	}
+	return &guardPb.ContractMeta{
+		ContractId:       shard.ContractID,
+		RenterPid:        session.Renter.Pretty(),
+		HostPid:          shard.Receiver.Pretty(),
+		ShardHash:        chunkHash,
+		ShardIndex:       chunkIndex,
+		FileHash:         session.FileHash.KeyString(),
+		RentStart:        shard.StartTime,
+		RentEnd:          shard.StartTime.Add(shard.Length),
+		GuardPid:         guardPid.Pretty(),
+		EscrowPid:        escrowPid.Pretty(),
+		Price:            shard.Price,
+		Amount:           1000,
+		CollateralAmount: 0,
+		PayoutSchedule:   0,
+		NumPayouts:       5,
+	}, nil
+}
+
+func SignedContractAndMarshal(meta *guardPb.ContractMeta, cont *guardPb.Contract, privKey ic.PrivKey, isPayer bool) ([]byte, error) {
+	sig, err := crypto.Sign(privKey, meta)
+	if err != nil {
+		return nil, err
+	}
+	if cont == nil {
+		cont = &guardPb.Contract{
+			ContractMeta:   *meta,
+			LastModifyTime: time.Now(),
+		}
+	} else {
+		cont.LastModifyTime = time.Now()
+	}
+	if isPayer {
+		cont.RenterSignature = sig
+	} else {
+		cont.HostSignature = sig
+	}
+	return proto.Marshal(cont)
+}
+
+func UnmarshalGuardContract(marshaledBody []byte) (*guardPb.Contract, error) {
+	signedContract := &guardPb.Contract{}
+	err := proto.Unmarshal(marshaledBody, signedContract)
+	if err != nil {
+		return nil, err
+	}
+	return signedContract, nil
+}
+
+func getGuardAndEscrowPid(configuration *config.Config) (peer.ID, peer.ID, error) {
+	escrowPid, err := pidFromString(configuration.Services.EscrowPubKeys[0])
+	if err != nil {
+		return "", "", err
+	}
+	guardPid, err := pidFromString(configuration.Services.GuardPubKeys[0])
+	if err != nil {
+		return "", "", err
+	}
+	return guardPid, escrowPid, err
+}
+
+// Todo: modify or change it all
+//func NewFileStoreStatus(session *storage.FileContracts, endTime time.Time, configuration *config.Config) (*guardPb.FileStoreStatus, error) {
+//
+//	escrowPid, err := pidFromString(configuration.Services.EscrowPubKeys[0])
+//	if err != nil {
+//		return nil, err
+//	}
+//	guardPid, err := pidFromString(configuration.Services.GuardPubKeys[0])
+//	if err != nil {
+//		return nil, err
+//	}
+//	fileStoreMeta := guardPb.FileStoreMeta{
+//		RenterPid:        session.Renter.Pretty(),
+//		FileHash:         session.FileHash.KeyString(),
+//		FileSize:         2000000000, // default??
+//		RentStart:        time.Now(),
+//		RentEnd:          endTime,
+//		CheckFrequency:   0,
+//		GuardFee:         0,
+//		EscrowFee:        0,
+//		ShardCount:       int32(len(session.ShardInfo)),
+//		MinimumShards:    10,
+//		RecoverThreshold: 20,
+//		EscrowPid:        escrowPid.Pretty(),
+//		GuardPid:         guardPid.Pretty(),
+//	}
+//
+//}
+
+func pidFromString(key string) (peer.ID, error) {
+	pubKey, err := crypto.ToPubKey(key)
```

Comment:
there is a function for this in go-btfs-common/crypto .. i think?

Reply:
no, but I can put it there